aten/src/ATen/native/cuda/UniqueCub.cu

```diff
@@ -0,0 +1,201 @@
+#include <ATen/native/cuda/UniqueCub.cuh>
+
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/cuda/CUDAApplyUtils.cuh>
+#include <ATen/cuda/cub.cuh>
+
+namespace at {
+namespace native {
+namespace internal {
+
+namespace {
+
+template <typename scalar_t>
+__global__ void adjacent_difference_kernel(
+    int64_t n,
+    const scalar_t* input,
+    int* output) {
+  auto begin = blockIdx.x * blockDim.x + threadIdx.x;
+  auto stride = gridDim.x * blockDim.x;
+  for (auto i = begin; i < n; i += stride) {
+    output[i] = i > 0 ? input[i] != input[i - 1] : 0;
+  }
+}
+
+__global__ void scatter_kernel(
+    int64_t n,
+    const int64_t* input,
+    const int64_t* indices,
+    int64_t* output) {
+  auto begin = blockIdx.x * blockDim.x + threadIdx.x;
+  auto stride = gridDim.x * blockDim.x;
+  for (auto i = begin; i < n; i += stride) {
+    output[indices[i]] = input[i];
+  }
+}
+
+// A variation of compute_unique (defined in Unique.cu) that doesn't allow
+// customizing equal and not_equal (needed to use CUB).
+template <typename scalar_t>
+std::tuple<Tensor, Tensor, Tensor, int64_t> compute_unique(
+    const Tensor& sorted,
+    const Tensor& sorted_indices,
+    const bool return_inverse,
+    const bool return_counts) {
+  int64_t num_inp = sorted.numel();
+  TORCH_CHECK(
+      num_inp <= INT_MAX, "num_inp ", num_inp, " is too big to for CUB");
+  auto options = sorted.options().dtype(kLong);
+  const scalar_t* data = sorted.data_ptr<scalar_t>();
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+
+  // inverse indices
+  Tensor inverse_indices;
+  if (!return_inverse || num_inp == 0) {
+    inverse_indices = at::empty({0}, options);
+  } else {
+    TORCH_CHECK(
+        sorted_indices.defined(),
+        "return_inverse is set to true, but sorted_indices is undefined. Send a bug report!");
+    inverse_indices = at::empty({num_inp}, options);
+    int* inv_loc_ptr =
+        reinterpret_cast<int*>(inverse_indices.data_ptr<int64_t>());
+    const dim3 block =
+        dim3(std::min(static_cast<int64_t>(cuda::getApplyBlock().x), num_inp));
+    dim3 grid;
+    int curDevice = -1;
+    cudaGetDevice(&curDevice);
+    cuda::getApplyGrid(num_inp, grid, curDevice);
+    adjacent_difference_kernel<scalar_t>
+        <<<grid, block, 0, stream>>>(num_inp, data, inv_loc_ptr);
+    C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+    size_t temp_storage_bytes = 0;
+    Tensor inv_loc_out = at::empty({num_inp}, options);
+    CUB_WRAPPER(
+        cub::DeviceScan::InclusiveSum,
+        inv_loc_ptr,
+        inv_loc_out.data_ptr<int64_t>(),
+        num_inp,
+        stream);
+
+    scatter_kernel<<<grid, block, 0, stream>>>(
+        num_inp,
+        inv_loc_out.data_ptr<int64_t>(),
+        sorted_indices.data_ptr<int64_t>(),
+        inverse_indices.data_ptr<int64_t>());
```

Comment:
Looks like for the case of `consecutive  == true`, `sorted_indices` will be `0, 1, 2, 3, ...`, therefore the `scatter_kernel` is effectively a memcpy? You can add an if statement for this case and make `InclusiveSum` directly output to `inverse_indices.data_ptr<int64_t>()`.